### ISSUE TITLE
postgres: Read blob/bytea as stream via LargeObject api

### ISSUE THREAD

**Issue by gpsfl** [jdbi/jdbi#1706]
How do you read a blob (or bytea in postgres to be exact) as a stream using jdbi?

With normal jdbc it would be `ResultSet#getBinaryStream`, however I couldn't find any equivalent in jdbi. I tried using `mapTo(InputStream.class)` and `mapTo(Blob.class)`, but both failed with a `NoSuchMapperException`.

It seems to work with `mapTo(byte[].class)` however I'd like to avoid loading all the data at once (due to memory/performance).

**Comment by leaumar** [comment]
I think supporting this would be a oneliner for us. I may be wrong though.
```java
mappers.put(InputStream.class, new GetterMapper<>(ResultSet::getBinaryStream));
```
Meanwhile, you could just register this on `ColumnMappers` in your config:
```java
public class InputStreamColumnMapper implements ColumnMapper<InputStream> {
    @Override
    public InputStream map(ResultSet r, int columnNumber, StatementContext ctx) throws SQLException {
        return r.getBinaryStream(columnNumber);
    }
}
```

**Comment by stevenschlansker** [comment]
Do note that if we do add such a mapper, there's a good chance the data is not properly streamed at all.  I think it will be transferred as a `byte[]` and then read through some analogue of `ByteArrayInputStream`.  Maybe that's fine for your use case.  But if you do actually need to stream, the setup is a bit more complicated: https://jdbc.postgresql.org/documentation/head/binary-data.html

**Comment by stevenschlansker** [comment]
@gpsfl , your problem statement is a bit contradictory: do you use a `bytea` or a `blob` column type?
If it's the former, I think the answer is that the driver cannot stream the data and you are best just retrieving as a `byte[]`.  If it's actually stored in a blob column, then this could be a feature request to support a facade over Postgres large object storage to make it easy to use in Jdbi.

**Comment by gpsfl** [comment]
As far as I know postgres does not have a blob type (large objects are `oid`s). To clarify: We are currently using a `bytea` in postgres. We now switched to using a byte[] in jdbi based on what you said about it the binary stream just being a ByteArrayInputStream. However it would be great to see some support for postgres large object as streaming should provide a better performance.

**Comment by stevenschlansker** [comment]
You're right, it seems Postgres doesn't have a blob type - but per jdbc docs, using the `oid` based large object storage is how you do streaming.  I agree, easy large object streaming is a good feature request.  Thanks for suggesting it!

**Comment by stevenschlansker** [comment]
I verified my theory.  Check out the implementation of `getBinaryStream()`:

```
    /**
     * @see java.sql.ResultSet#getBinaryStream(int)
     */
    public InputStream getBinaryStream(int col) throws SQLException {
        return new ByteArrayInputStream(getBytes(col));
    }
```

**Comment by stevenschlansker** [comment]
One question: how does cleanup work?  It looks like LargeObjects must be explicitly deleted.  But of course, when you run a SQL delete statement that removes a row containing an oid pointing at the object, users will expect it to go away.

**Comment by stevenschlansker** [comment]
Hi @gpsfl, it's not done yet, but does this look like it might help you? https://github.com/jdbi/jdbi/pull/1711

**Comment by stevenschlansker** [comment]
I am able to store 256MB data blobs with only 96m of Java heap :)

**Comment by gpsfl** [comment]
`the code is now ready. If you have time to review and test it that'd be much appreciated.`
What do do exactly? I can test if it works tho I don't think I have enough knowledge on that to do any meaningful code review. Anything to watch out for when build it or does `mvn install` just do the job?

**Comment by stevenschlansker** [comment]
Yup, now that it's merged into master, you can just build jdbi `3.14.0-SNAPSHOT` and test it with your project if you like.  I'm closing this issue since it's merged, but please feel free to follow up with any issues or questions.
